tensorflow/python/ops/image_ops_impl.py

```diff
@@ -1515,6 +1517,7 @@ def decode_image(contents, channels=None, name=None):
     if channels not in (None, 0, 1, 3, 4):
       raise ValueError('channels must be in (None, 0, 1, 3, 4)')
     substr = string_ops.substr(contents, 0, 3)
+    dtype = dtypes.uint8 if dtype == dtypes.uint8 else dtypes.uint16
```

Comment:
Oh, not here. This should be in the call to `decode_png`, right?

Reply:
In function tf.cond(pred, fn1, fn2), fn1 and fn2 must have the same type,
other formats decode function also need to returned uint16 when dtype != uint8.

Reply:
Perhaps I should add convert_image_dtype in decode_{jpeg, png, gif, bmp}, to make sure that returned tensors have same type.

Something like:
```python
def _png():
  ...
  return convert_image_dtype(gen_image_ops.decode_png(contents, channels,
    dtype=dtypes.uint8 if dtype==dtypes.uint8 else dtypes.uint16), dtype)

def _jpeg():
  ...
  return convert_image_dtype(gen_image_ops.decode_jpeg(contents, channels), dtype)
```